### PR TITLE
implemented possibility of chosing random seed

### PR DESCRIPTION
**Pull request by marekyggdrasil**

### CODE REVIEW

**Comment by Ericgig** [comment]
Thank you Marek,
It seems you forgot to add a seed variable in the definition of the rand_herm function (line 112~114). Because of this, many tests fails. 
Could you add it?

**Comment by marekyggdrasil** [comment]
@Ericgig sorry about that, I just added the argument, I'm also planning adding some tests but I haven't yet figured out what would be best way to do it, any suggestions welcome!

**Comment by marekyggdrasil** [comment]
How about instead of checking if it is Hermitian, we could put a seed parameter and check if the output is exactly equal (or [close](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html)) to given hardcoded matrix?

https://github.com/qutip/qutip/blob/d91a4c93f35d97b8bfc633420ffa6ddebadc06a1/qutip/tests/test_rand.py#L57

**Comment by Ericgig** [comment]
Hardcoded matrix will cause problem in later modification of the code. (Or if numpy update it's random number generator, etc.)
Keeping the present tests, I would add something like this:
```
O1 = rand_herm(..., seed=seed)
O2 = rand_herm(..., seed=None)
O3 = rand_herm(..., seed=seed)
assert(O1!=O2)
assert(O1==O3)
```
You check that random Qobj are reproducible using a seed and work as expected when not.

**Comment by marekyggdrasil** [comment]
@Ericgig good idea about the tests, I'll implement them in this way

now there seems to be some problem with continuous integration, it fails on

```
ERROR: Failure: ImportError (libgfortran.so.1: cannot open shared object file: No such file or directory)
```

I attempted to fix it by installing `libgfortran` using `conda`, however it didn't fix the issue

https://github.com/travis-ci/travis-ci/issues/4816#issuecomment-148789525
https://github.com/menpo/landmarkerio-server/issues/23#issuecomment-278361601

I don't want to mess up too much with continous integration as it takes long time to rebuild it, any suggestions on that?

**Comment by Ericgig** [comment]
Do worry about the failing python2.7 tests, it's a common error on all recent checks. Your pull should not affect them.

**Comment by marekyggdrasil** [comment]
@Ericgig tests implemented and passing, please code review if you can.

**Comment by Ericgig** [comment]
As it is in your code, if no seed is given, then np.random.seed(None) is called. np.random.seed(None) reset the seed based on the computer time. 
It would break the reproducibility of code which already set the seed once:
```
np.random.seed(seed)
my_rand_qobj = qt.rand_herm(N)
```

It would be better if you did:
```
if seed is not None:
    np.random.seed(seed=seed)
```

**Comment by marekyggdrasil** [comment]
@Ericgig thanks for suggestion

weird, my local tests all passed, let me examine the problem

**Comment by Ericgig** [comment]
It seems to be caused by a new version of cython which affect which files are used for tests by nose. I made a patch.

**Comment by marekyggdrasil** [comment]
Thats great, thanks @Ericgig 

If any other modifications requested don't hesitate to let me know. Looking forward to have my changes merged!

File: qutip/random_objects.py [hunk 1]
```diff
@@ -150,8 +154,10 @@ def rand_herm(N, density=0.75, dims=None, pos_def=False):
             _check_dims(dims, N, N)
         nvals = N**2*density
         while M.nnz < 0.95*nvals:
-            M = rand_jacobi_rotation(M)
+            M = rand_jacobi_rotation(M, seed=seed)
```
Comment (Ericgig) [inline qutip/random_objects.py]: Putting the seed here will reseed at each iteration of the while loop.
Better to to it just once at the beginning of the function.

File: qutip/random_objects.py [hunk 2]
```diff
@@ -352,7 +360,7 @@ def rand_dm(N, density=0.75, pure=False, dims=None):
             _check_dims(dims, N, N)
         nvals = N**2*density
         while H.nnz < 0.95*nvals:
-            H = rand_jacobi_rotation(H)
+            H = rand_jacobi_rotation(H, seed=seed)
```
Comment (Ericgig) [inline qutip/random_objects.py]: Here the code is also calling seed inside a loop.

File: qutip/random_objects.py [hunk 3]
```diff
@@ -494,15 +502,17 @@ def rand_super(N=5, dims=None):
         pass
     else:
         dims = [[[N],[N]], [[N],[N]]]
-    H = rand_herm(N)
+    H = rand_herm(N, seed=seed)
+    if seed is not None:
```
Comment (Ericgig) [inline qutip/random_objects.py]: The seed is already applied in rand_herm, no need to do it twice.

**Comment by Ericgig** [comment]
I marked a few spot where the seed was applied more than once.
The failed test is a known issue.

**Comment by marekyggdrasil** [comment]
@Ericgig I marked your remarks as resolved and made some new changes. If anything more required don't hesitate to ping me.

**Pull request merged**